### PR TITLE
Perform unreferenced file cleanup for any operation failure and mute flaky test

### PR DESCRIPTION
**Pull request by RS146BIJAY**
### Description
As of now, we cleanup unreferenced files whenever last write is performed by [merge ](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/Engine.java#L1370)and it caused disk to get full and shard to fail. Incase some other operation performs the last write and caused disk to get 100% full and a merge is ongoing, merge will just get aborted and no cleanup will be performed. Since when closing the shard it is the other operation causing disk full and not segment merge.  

In order to fix this we need to cleanup unreferenced files when any operation failed due to disk full.

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/12054

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/Engine.java [hunk 1]
```diff
@@ -1317,11 +1317,11 @@ public void failEngine(String reason, @Nullable Exception failure) {
...
-                    // If cleanup of unreferenced flag is enabled and force merge or regular merge failed due to IOException,
-                    // clean all unreferenced files on best effort basis created during failed merge and reset the
-                    // shard state back to last Lucene Commit.
-                    if (shouldCleanupUnreferencedFiles() && isMergeFailureDueToIOException(failure, reason)) {
-                        logger.info("Cleaning up unreferenced files as merge failed due to: {}", reason);
+                    // If cleanup of unreferenced file flag is enabled and any operation failed due to IOException,
+                    // clean up all unreferenced files on best effort basis created during failed merge and reset the
+                    // shard state back to last Lucene Commit
+                    if (shouldCleanupUnreferencedFiles() && isOperationFailureDueToIOException(failure)) {
+                        logger.info("Cleaning up unreferenced files created during failed merge due to: {}", reason);
```
Comment (deshsidd) [inline server/src/main/java/org/opensearch/index/engine/Engine.java:1324]: Should be warning?

Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/engine/Engine.java:1324]: We would be doing extra clean ups on may be other shards as well due to this . Please evaluate if we can avoid the same and do clean up only on the responsible shard instead

Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/engine/Engine.java:1324]: Do we need to log if the exception isn't IOE

**Comment by gbbafna** [comment]
@RS146BIJAY  : are you still working on this change ?

**Pull request closed without merging**